streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -182,30 +193,49 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
                 cleanupTask(task);
 
                 try {
-                    task.closeClean();
+                    checkpointPerTask.put(task, task.prepareCloseClean());
+                    if (task.state() != CREATED) {
+                        consumedOffsetsAndMetadataPerTask.put(task.id(), task.committableOffsetsAndMetadata());
+                    }
                 } catch (final RuntimeException e) {
                     final String uncleanMessage = String.format("Failed to close task %s cleanly. Attempting to close remaining tasks before re-throwing:", task.id());
                     log.error(uncleanMessage, e);
                     taskCloseExceptions.put(task.id(), e);
                     // We've already recorded the exception (which is the point of clean).
                     // Now, we should go ahead and complete the close because a half-closed task is no good to anyone.
-                    task.closeDirty();
-                } finally {
-                    if (task.isActive()) {
-                        try {
-                            activeTaskCreator.closeAndRemoveTaskProducerIfNeeded(task.id());
-                        } catch (final RuntimeException e) {
-                            final String uncleanMessage = String.format("Failed to close task %s cleanly. Attempting to close remaining tasks before re-throwing:", task.id());
-                            log.error(uncleanMessage, e);
-                            taskCloseExceptions.putIfAbsent(task.id(), e);
-                        }
-                    }
+                    task.prepareCloseDirty();
+                    dirtyTasks.add(task);
                 }
 
                 iterator.remove();
             }
         }
 
+        if (!consumedOffsetsAndMetadataPerTask.isEmpty()) {
+            commitOffsetsOrTransaction(consumedOffsetsAndMetadataPerTask);
+        }
+
+        for (final Map.Entry<Task, Map<TopicPartition, Long>> taskAndCheckpoint : checkpointPerTask.entrySet()) {
+            final Task task = taskAndCheckpoint.getKey();
+            try {
+                task.closeClean(checkpointPerTask.get(task));
+            } catch (final RuntimeException e) {
```

Comment:
We lack unit test coverage for this case

Reply:
I know...